src/machine.rs

```diff
@@ -222,124 +222,89 @@ impl<R: Register, M: Memory> SupportMachine for DefaultCoreMachine<R, M> {
-            registers: Default::default(),
-            pc: R::zero(),
-            memory: M::default(),
-            elf_end: 0,
-            cycles: 0,
-            max_cycles: None,
-        }
-    }
-}
-
-impl<R, M> DefaultCoreMachine<R, M>
-where
-    Self: Default,
-{
-    pub fn new_with_max_cycles(max_cycles: u64) -> DefaultCoreMachine<R, M> {
+impl<R: Register, M: Memory> DefaultCoreMachine<R, M> {
+    pub fn new_with_max_cycles(max_cycles: u64) -> Self {
         Self {
             max_cycles: Some(max_cycles),
-            ..Self::default()
+            ..Default::default()
         }
     }
 }
 
 pub type InstructionCycleFunc = Fn(&Instruction) -> u64;
 
-pub struct DefaultMachine<'a, R, M> {
-    core: DefaultCoreMachine<R, M>,
+#[derive(Default)]
+pub struct DefaultMachine<'a, Core> {
+    core: Core,
 
     // We have run benchmarks on secp256k1 verification, the performance
     // cost of the Box wrapper here is neglectable, hence we are sticking
     // with Box solution for simplicity now. Later if this becomes an issue,
     // we can change to static dispatch.
     instruction_cycle_func: Option<Box<InstructionCycleFunc>>,
-    syscalls: Vec<Box<dyn Syscalls<R, M> + 'a>>,
+    syscalls: Vec<Box<dyn Syscalls<Core> + 'a>>,
     running: bool,
     exit_code: u8,
 }
 
-impl<'a, R, M> Deref for DefaultMachine<'a, R, M> {
-    type Target = DefaultCoreMachine<R, M>;
+impl<Core: CoreMachine> CoreMachine for DefaultMachine<'_, Core> {
+    type REG = <Core as CoreMachine>::REG;
+    type MEM = <Core as CoreMachine>::MEM;
 
-    fn deref(&self) -> &Self::Target {
-        &self.core
-    }
-}
-
-impl<'a, R, M> DerefMut for DefaultMachine<'a, R, M> {
-    fn deref_mut(&mut self) -> &mut DefaultCoreMachine<R, M> {
-        &mut self.core
-    }
-}
-
-impl<'a, R: Register, M: Memory> CoreMachine for DefaultMachine<'a, R, M> {
-    type REG = R;
-    type MEM = M;
-    fn pc(&self) -> &R {
-        &self.pc
+    fn pc(&self) -> &Self::REG {
+        &self.core.pc()
     }
 
-    fn set_pc(&mut self, next_pc: R) {
-        self.pc = next_pc;
+    fn set_pc(&mut self, next_pc: Self::REG) {
+        self.core.set_pc(next_pc)
     }
 
-    fn memory(&self) -> &M {
-        &self.memory
+    fn memory(&self) -> &Self::MEM {
+        self.core.memory()
     }
 
-    fn memory_mut(&mut self) -> &mut M {
-        &mut self.memory
+    fn memory_mut(&mut self) -> &mut Self::MEM {
+        self.core.memory_mut()
     }
 
-    fn registers(&self) -> &[R] {
-        &self.registers
+    fn registers(&self) -> &[Self::REG] {
+        self.core.registers()
     }
 
-    fn set_register(&mut self, idx: usize, value: R) {
-        self.registers[idx] = value;
+    fn set_register(&mut self, idx: usize, value: Self::REG) {
+        self.core.set_register(idx, value)
     }
 }
 
-impl<'a, R: Register, M: Memory> SupportMachine for DefaultMachine<'a, R, M> {
+impl<Core: SupportMachine> SupportMachine for DefaultMachine<'_, Core> {
```

Comment:
Can we rename `Core` here to something else, maybe `Mac`? It feels a little weird that `Core` implements `SupportMachine`, while we also have `CoreMachine`